media/2021-05-14-paying-homage-to-the-compall-michael-pattern/5-scala3-opaque-types.worksheet.sc

```diff
@@ -0,0 +1,47 @@
+// Approach using Scala 3 Opaque Type Aliases
+
+object Units {
+  opaque type Kilometers = Double
+  opaque type Miles = Double
+
+  val ZeroKm: Kilometers = 0
+  val ZeroMi: Miles = 0
+
+  def kilometres(value: Double): Option[Kilometers] = if (value < 0) None else Some(value)
+  def miles(value: Double): Option[Miles] = if (value < 0) None else Some(value)
+
+  def add(km1: Kilometers, km2: Kilometers): Kilometers = km1 + km2
+  def toKilometers(miles: Miles): Kilometers = miles * 1.6
+
+  extension (km: Kilometers) {
+    def + (km2: Kilometers): Kilometers = add(km, km2)
+  }
+
+  extension (miles: Miles) {
+    def toKm: Kilometers = toKilometers(miles)
+  }
+}
+
+import Units._
+
+class Booster() {
+  def provideLaunchBoost(): Miles = miles(100).getOrElse(ZeroMi)
+}
+
+class Rocket(booster: Booster) {
+  private var distance: Kilometers = ZeroKm
+
+  def launch(): Unit = {
+    // Kilometers and Miles are different types. So compiler prevents the previous bug
+    distance += booster.provideLaunchBoost().toKm
+  }
+
+  def distanceTravelled: Kilometers = distance
+}
+
+// For fun, let's make use of Scala 3's Universal Apply Methods to omit the 'new'
+val rocket: Rocket = Rocket(Booster())
```

Comment:
@ckipp01 to the point about `Booster` potentially being a case class from the beginning, I actually had in mind that Booster would be stateful (in which case `case class` wouldn't be a good fit). I started down a road to add fuel to the Booster which we would mutate after each `provideLaunchBoost`, but then that quickly got out of hand because to be consistent with the message of the article the units for the fuel (`Gallons` or `Litres`) would also need to be modelled as opaque types. 😢

Reply:
Ha, sort of assumed there was a reason like this behind the scenes.

Reply:
just thought of another way 🤔 . I'll try that and see if it doesn't make things too complicated

Reply:
no, too complicated. I tried adding a mutable variable "launchCannisters" inside Booster that would be decremented each time. but then that makes it a bit noisy and distracts from the main message.